sdk/core/azure-core-tracing-opentelemetry/src/main/java/com/azure/core/tracing/opentelemetry/OpenTelemetryTracer.java

```diff
@@ -0,0 +1,369 @@
+        Objects.requireNonNull(processKind, "'processKind' cannot be null.");
+
+        Span span;
+        Builder spanBuilder;
+
+        switch (processKind) {
+            case SEND:
+                spanBuilder = getSpanBuilder(spanName, context);
+                span = spanBuilder.setSpanKind(Span.Kind.PRODUCER).startSpan();
+                if (span.isRecording()) {
+                    // If span is sampled in, add additional request attributes
+                    addSpanRequestAttributes(span, context, spanName);
+                }
+                return context.addData(PARENT_SPAN_KEY, span);
+            case MESSAGE:
+                spanBuilder = getSpanBuilder(spanName, context);
+                span = spanBuilder.startSpan();
+                // Add diagnostic Id and trace-headers to Context
+                context = setContextData(span);
+                return context.addData(PARENT_SPAN_KEY, span);
+            case PROCESS:
+                return startScopedSpan(spanName, context);
+            default:
+                return Context.NONE;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void end(int responseCode, Throwable throwable, Context context) {
+        Objects.requireNonNull(context, "'context' cannot be null.");
+        final Span span = getSpan(context);
+        if (span == null) {
+            return;
+        }
+
+        if (span.isRecording()) {
+            span.setStatus(HttpTraceUtil.parseResponseStatus(responseCode, throwable));
+        }
+
+        span.end();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setAttribute(String key, String value, Context context) {
+        if (CoreUtils.isNullOrEmpty(value)) {
+            logger.warning("Failed to set span attribute since value is null or empty.");
+            return;
+        }
+
+        final Span span = getSpan(context);
+        if (span != null) {
+            span.setAttribute(key, AttributeValue.stringAttributeValue(value));
+        } else {
+            logger.warning("Failed to find span to add attribute.");
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Context setSpanName(String spanName, Context context) {
+        return context.addData(USER_SPAN_NAME_KEY, spanName);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void end(String statusMessage, Throwable throwable, Context context) {
+        final Span span = getSpan(context);
+        if (span == null) {
+            logger.warning("Failed to find span to end it.");
+            return;
+        }
+
+        if (span.isRecording()) {
+            span.setStatus(AmqpTraceUtil.parseStatusMessage(statusMessage, throwable));
+        }
+
+        span.end();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void addLink(Context context) {
+        // TODO: Add links not working as expected, confirm this
```

Comment:
is  this comment still relevant?

Reply:
yes, will track this here - https://github.com/Azure/azure-sdk-for-java/issues/6188